sys/vm_map.c

```diff
@@ -181,6 +181,25 @@ int vm_map_findspace(vm_map_t *map, vm_addr_t start, size_t length,
   return 0;
 }
 
+int vm_map_resize(vm_map_t *map, vm_map_entry_t *entry, vm_addr_t new_end) {
+  assert(is_aligned(new_end, PAGESIZE));
+  if (new_end > entry->end) {
+    /* Expanding entry */
+    vm_map_entry_t *next = TAILQ_NEXT(entry, map_list);
+    vm_addr_t gap_end = next ? next->start : map->pmap->end;
+    if (new_end > gap_end)
+      return -ENOMEM;
+  } else {
+    /* Shrinking entry */
+    if (new_end < entry->start)
+      return -ENOMEM;
+    /* TODO: Invalidate tlb? */
```

Comment:
Certainly we have to unmap and free pages that are in released region. `pmap` interface should deliver abstractions that do not require explicit operations on TLB. However `pmap` is  incomplete in many aspects.